services/types.go

```diff
@@ -69,49 +90,137 @@ type ServiceSharding interface {
+type MetadataService interface {
+	// Metadata returns the metadata for a given service
+	Metadata() (ServiceMetadata, error)
 
-	// Query returns metadata and a list of available instances for a given service
-	Query(service string, opts QueryOptions) (Service, error)
+	// SetMetadata sets the metadata for a given service
+	SetMetadata(m ServiceMetadata) error
+}
 
-	// Watch returns a watch on metadata and a list of available instances for a given service
-	Watch(service string, opts QueryOptions) (xwatch.Watch, error)
+// PlacementService handles the placement related operations for registered services
+// all write or update operations will persist the generated placement before returning success
+type PlacementService interface {
+	// BuildInitialPlacement initialize a placement
+	BuildInitialPlacement(instances []PlacementInstance, numShards int, rf int, popts PlacementOptions) (ServicePlacement, error)
+
+	// AddReplica up the replica factor by 1 in the placement
+	AddReplica(popts PlacementOptions) (ServicePlacement, error)
+
+	// AddInstance picks an instance from the candidate list to the placement
+	AddInstance(candidates []PlacementInstance, popts PlacementOptions) (ServicePlacement, error)
+
+	// RemoveInstance removes an instance from the placement
+	RemoveInstance(i PlacementInstance, popts PlacementOptions) (ServicePlacement, error)
+
+	// ReplaceInstance picks instances from the candidate list to replace an instance in current placement
+	ReplaceInstance(leavingInstance PlacementInstance, candidates []PlacementInstance, popts PlacementOptions) (ServicePlacement, error)
+
+	// Placement gets the persisted placement for service
+	Placement() (ServicePlacement, error)
+}
+
+// PlacementOptions is the interface for placement options
+type PlacementOptions interface {
+	// LooseRackCheck enables the placement to loose the rack check
+	// during instance replacement to achieve full ownership transfer
+	LooseRackCheck() bool
+	SetLooseRackCheck(looseRackCheck bool) PlacementOptions
+
+	// AllowPartialReplace allows shards from the leaving instance to be
+	// placed on instances other than the new instances in a replace operation
+	AllowPartialReplace() bool
+	SetAllowPartialReplace(allowPartialReplace bool) PlacementOptions
+}
+
+// ServicePlacement describes how instances are placed in a service
+type ServicePlacement interface {
+	// Instances returns all Instances in the placement
+	Instances() []PlacementInstance
+
+	// NumInstances returns the number of instances in the placement
+	NumInstances() int
+
+	// Instance returns the Instance for the requested id
+	Instance(id string) PlacementInstance
+
+	// ReplicaFactor returns the replica factor in the placement
+	ReplicaFactor() int
+
+	// Shards returns all the unique shard ids for a replica
+	Shards() []uint32
+
+	// ShardsLen returns the number of shards in a replica
+	NumShards() int
+
+	// Validate checks if the ServicePlacement is valid
+	Validate() error
```

Comment:
Check if `Validate` and `Copy` are only used inside the placement service and if so, move them inside `PlacementService`